systemd: allow custom unit folders to be configured with SYSTEMD_UNIT…

svanderburg opened the pull request:
…_PATH

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

This feature makes it possible to create mutable unit directories for systemd, in arbitrary locations. This is useful for experimentation purposes, and for integration with tools that require a mutable unit directory whose contents is stored persistently.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

svanderburg committed:
systemd: allow custom unit folders to be configured with SYSTEMD_UNIT_PATH

cole-h commented on nixos/modules/system/boot/stage-2.nix:
```diff
@@ -20,6 +20,11 @@ let
+    systemdUnitPathEnvVar = lib.optionalString (config.boot.extraSystemdUnitPaths != [])
+      ("SYSTEMD_UNIT_PATH="
+      + builtins.concatStringsSep ":" config.boot.extraSystemdUnitPaths
+      + ":");
```
The trailing `:` confused me briefly before looking it up. Would it be prudent to document the reason it's necessary?

From the `systemd.unit` manpage: "If `$SYSTEMD_UNIT_PATH` ends with an empty component ("`:`"), the usual unit load path will be appended to the contents of the variable."

cole-h commented on nixos/modules/system/boot/stage-2.nix:
```diff
@@ -20,6 +20,11 @@ let
       pkgs.util-linux
     ] ++ lib.optional useHostResolvConf pkgs.openresolv);
     fsPackagesPath = lib.makeBinPath config.system.fsPackages;
+    inherit (config.boot) extraSystemdUnitPaths;
```
Maybe add this to the above `inherit (config.boot) systemdExecutable` instead of having a separate one?

svanderburg commented:
@cole-h Good points. I just implemented your suggestions.

SuperSandro2000 commented on nixos/modules/system/boot/stage-2.nix:
```diff
@@ -82,8 +86,18 @@ in
+      };
     };
 
+
```
Please revert that.

SuperSandro2000 commented on nixos/tests/systemd-unit-path.nix:
```diff
@@ -0,0 +1,46 @@
+  exampleScript = pkgs.writeTextFile {
+    name = "example.sh";
+    text = ''
+      #! ${pkgs.stdenv.shell} -e
+
+      while true; do
+          echo "Example script running" >&2
+          ${pkgs.coreutils}/bin/sleep 1
```
Is this cross compile friendly?

svanderburg commented:
@SuperSandro2000 I just made a correction.

I didn't know that test cases were also supposed to be cross compilable.

svanderburg commented:
@aanderse @arianvp Could you also give this a try and see if it works for you?

svanderburg commented:
Any objections integrating this? I have the privileges to do merge this, if nobody objects.

I have implemented all the feedback given so far. Furthermore, I've been using it for my own use cases already for several days and seems to work just fine.

aanderse reviewed (approved):
This change is great :+1: Big thanks to @arianvp for finding such a nice solution, bigger thanks to @svanderburg for creating the PR :tada: 

I haven't tested this exact PR, but this is right inline with the testing I was doing earlier.

LGTM - please merge if no one has any issues.

svanderburg commented:
If nobody has any objections, then I'll merge it tonight (CET timezone)

arianvp commented on nixos/modules/system/boot/stage-2-init.sh:
```diff
@@ -165,8 +165,16 @@ exec 1>&$logOutFd 2>&$logErrFd
 
+# Create directories for the extra systemd unit paths, so that they can be
+# used. A dangling directory in SYSTEMD_UNIT_PATH gets ignored by systemd.
+for p in @extraSystemdUnitPaths@; do
```
Hmm this is the only thing I'm not overly fond of. I don't think we should be creating these arbitrary directories in stage-2.  Especially if people make typos etc.

svanderburg replied:
I need to double check this, but what I noticed is that if this directory does not exists, then systemd ignores it. So if after booting this directory gets created, it will not work until the next reboot.

So upgrading from an old systemd release then requires you to reboot to get the `SYSTEMD_UNIT_PATH` setting working, and then another after the directory gets created.

arianvp commented:
Left one remark. apart from that lgtm.

AndersonTorres reviewed (approved):
LGTM

svanderburg commented:
@arianvp I just did some more tests and it turns out that you were right -- it seems that it works without the presence of that directory. Maybe I couldn't catch this properly in my manual test.

I have just removed the code from the `stage-2-init.sh` script and adjusted the test case to create the directory after the system was started. It still works.

arianvp commented:
Awesome. Lgtm then!

On Mon, 22 Mar 2021, 20:44 Sander van der Burg, ***@***.***>
wrote:

> @arianvp <https://github.com/arianvp> I just did some more tests and it
> turns out that you were right -- it seems that it works without the
> presence of that directory. Maybe I couldn't catch this properly in my
> manual test.
>
> I have just removed the code from the stage-2-init.sh script and adjusted
> the test case to create the directory after the system was started. It
> still works.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/116455#issuecomment-804345158>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEZNI5MQODQA3IECCB3WMDTE6M23ANCNFSM4ZHDGPAQ>
> .
>

arianvp commented:
@aanderse in systemd 248 we'll be able to change the environment without reboot. From the release notes:

The environment block of the manager itself may be configured through
          a new ManagerEnvironment= setting in system.conf or user.conf. This
          complements existing ways to set the environment block (the kernel
          command line for the system manager, the inherited environment and
          user@.service unit file settings for the user manager).

aanderse commented:
@arianvp that's great to hear. a thanks for mentioning.
